Do not enumerate devices not initialized by udev

benzea opened the pull request:
Doing that can cause weird issues with permissions. One example is fprintd trying to open the fingerprint reader but selinux has not yet been tagged and it is not allowed to do so. Other scenarios can also be possible as it is not uncommon to change the user/group of devices using udev.

This should fix it, unfortunately the umockdev testing is not really optimal …

EDIT: Downstream bug report is https://bugzilla.redhat.com/show_bug.cgi?id=2346771

benzea committed:
tests/umockdev: remove UMOCKDEV_HOTPLUG workaround

umockdev 0.17.7 is old enough to rely on it for testing. Remove the
workaround as it might not be compatible with the follow up changes.

Signed-off-by: Benjamin Berg <benjamin@sipsolutions.net>

benzea committed:
linux_udev: only return initialized devices during enumeration

The udev enumerator can return devices for which udev has not finished
processing its rules. As udev has not finished handling its rules, the
application may not yet be permitted to open the device.

For these devices a hotplug event will happen later on when udev sends
the "add" event.

Change the enumerator to only return initialized devices when udev is
running. There is no nice API to check whether this is the case, so just
do the same access() call that udev_queue_get_udev_is_active() would
also do.

Also add a umockdev test for this specific corner case. Note that other
tests need to be updated to work around poor umockdev emulation of the
udev behaviour. umockdev never writes udev database entries to its
/run/udev/data, which means that devices are only considered initialized
when an event has been received.

Signed-off-by: Benjamin Berg <benjamin@sipsolutions.net>

benzea commented on tests/umockdev.c:
```diff
@@ -1110,6 +1135,37 @@ test_hotplug_add_remove(UMockdevTestbedFixture * fixture, UNUSED_DATA)
 }
 
+static void
+test_hotplug_udev_not_ready(UMockdevTestbedFixture * fixture, UNUSED_DATA)
```
Dunno, a test is nice, but then maybe having one that potentially breaks if umockdev is fixed isn't …

I opened https://github.com/martinpitt/umockdev/issues/267

3v1n0 replied:
> Dunno, a test is nice, but then maybe having one that potentially breaks if umockdev is fixed isn't …
> 
> I opened https://github.com/martinpitt/umockdev/issues/267

Yeah... Probably it makes sense to keep it in another PR that lands when umockdev is fixed?

benzea commented on tests/umockdev.c:
```diff
@@ -409,6 +409,8 @@ test_fixture_add_canon(UMockdevTestbedFixture * fixture)
 		"E: DRIVER=usb\n"
 		"E: BUSNUM=001\n"
 		"E: DEVNUM=001\n"
+		"E: MAJOR=189\n"
```
Just FYI, this is the correct thing here and it is needed because the `is_initialized` check in the enumerator is special and considers any device without a device number always as initialized (`udev_device_get_is_initialized` behaves differently there).

mcuee commented:
Looks like the intention of this PR is good. But I am not able to review the implementation properly.

benzea commented:
> Looks like the intention of this PR is good. But I am not able to review the implementation properly.

I don't think it is that complicated. Most of the PR is adjusting the tests, but maybe you can ignore that? Either way, please ping me if you need someone else to look over it. I think that it would be nice if this makes it into the next release.

tormodvolden commented:
This sounds like something we want, and I think I have heard about issues caused by such "half-ready" devices. It seems from the downstream report this have had minimal to none testing though. Have you patched libusb in RedHat with this already?

BTW, you're saying the device will now be detected later on the udev "add" event. So the enumerator will (originally) also return devices for which there has been no udev add event signaled yet? And the add event coincides with "initialized" becoming true?

And I am fine with bumping umockdev requirements to the 3 year old 17.7 (available in the oldest supported Ubuntu LTS release as one data point).

benzea commented:
> This sounds like something we want, and I think I have heard about issues caused by such "half-ready" devices. It seems from the downstream report this have had minimal to none testing though. Have you patched libusb in RedHat with this already?

I am not with Red Hat anymore. I suppose I could submit a PR for the Fedora libusb package and try getting it in there.

> BTW, you're saying the device will now be detected later on the udev "add" event. So the enumerator will (originally) also return devices for which there has been no udev add event signaled yet? And the add event coincides with "initialized" becoming true?

Yup, exactly. The `enumerator` looks at sysfs, so it sees all devices known to the kernel. However, once udev is done, it writes a file into `/run/udev/data/` and fires the `"add"` event to notify all processes.

What the change effectively does is to check whether the file in `/run/udev/data/` has already been written. If it does not exist, then we know that an `"add"` signal will arrive in the future.

mcuee commented:
Does this PR affects those who build libusb without udev support (eg: relying on netlink)?

benzea committed:
linux_udev: add workaround for udev rules

Some applications/libraries try to discover USB devices from within udev
rules. This is wrong, as the device is not initialized by udev and
should not be discovered by libusb. The libusb device discovery should
be disabled within a udev rule and libusb_wrap_sys_device should be used
instead.

Add a workaround by reading the environment variables and detecting when
we seem to be running within a udev rule as the corresponding
environment variables are set. In that case, let the application
discover the uninitialized device if it is the one we are running for.

Signed-off-by: Benjamin Berg <benjamin@sipsolutions.net>

benzea commented:
My attempt at a workaround for the libmtp mtp-probe binary behaviour. Haven't really tested it yet, but one can see the general approach. I added the checks for the other environment variables just to be a bit more sure that we are in a udev environment, but that is probably overkill.
